vm-allocator/src/address.rs

```diff
@@ -0,0 +1,365 @@
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE-BSD-3-Clause file.
+//
+// SPDX-License-Identifier: Apache-2.0 AND BSD-3-Clause
+
+use std::collections::btree_map::BTreeMap;
+use std::fmt::{self, Display};
+use std::result;
+use vm_memory::{Address, GuestAddress, GuestUsize};
+
+#[derive(Debug)]
+pub enum Error {
+    Overflow,
+    Overlap,
+    UnalignedAddress,
+    NullRequest,
+}
+
+impl Display for Error {
+    // This trait requires `fmt` with this exact signature.
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use self::Error::*;
+
+        match self {
+            Overflow => write!(f, "Address being allocated is overflow"),
+            Overlap => write!(f, "Address being allocated is overlap"),
+            UnalignedAddress => write!(f, "Address being allocated is unaligned"),
+            NullRequest => write!(f, "Address being allocated is null"),
+        }
+    }
+}
+
+pub type Result<T> = result::Result<T, Error>;
+
+/// Manages allocating address ranges.
+/// Use `AddressAllocator` whenever an address range needs to be allocated to different users.
+///
+/// # Examples
+///
+/// ```
+/// # use vm_allocator::AddressAllocator;
+/// # use vm_memory::{Address, GuestAddress, GuestUsize};
+///   AddressAllocator::new(GuestAddress(0x1000), 0x10000, Some(0x100)).map(|mut pool| {
+///       assert_eq!(pool.allocate(None, 0x110), Some(GuestAddress(0x10e00)));
+///       assert_eq!(pool.allocate(None, 0x100), Some(GuestAddress(0x10c00)));
+///   });
+/// ```
+#[derive(Debug, Eq, PartialEq)]
+pub struct AddressAllocator {
+    base: GuestAddress,
+    end: GuestAddress,
+    alignment: GuestUsize,
+    ranges: BTreeMap<GuestAddress, GuestUsize>,
+}
+
+impl AddressAllocator {
+    /// Creates a new `AddressAllocator` for managing a range of addresses.
+    /// Can return `None` if `pool_base` + `pool_size` overflows a u64 or if alignment isn't a power
+    /// of two.
+    ///
+    /// * `pool_base` - The starting address of the range to manage.
+    /// * `pool_size` - The size of the address range in bytes.
+    /// * `align_size` - The minimum size of an address region to align to, defaults to four.
+    pub fn new(
+        base: GuestAddress,
+        size: GuestUsize,
+        align_size: Option<GuestUsize>,
+    ) -> Option<Self> {
+        if size == 0 {
+            return None;
+        }
+
+        let end = base.checked_add(size - 1)?;
+        let alignment = align_size.unwrap_or(4);
+        if !alignment.is_power_of_two() || alignment == 0 {
+            return None;
+        }
+
+        let mut allocator = AddressAllocator {
+            base,
+            end,
+            alignment,
+            ranges: BTreeMap::new(),
+        };
+
+        // Insert the last address as a zero size range.
+        // This is our end of address space marker.
+        allocator.ranges.insert(base.checked_add(size)?, 0);
```

Comment:
If we want to keep the `end` being inclusive, I see another place needs to call checked_sub(1) once we define `end` like that :)
```
        let mut allocator = AddressAllocator {
            base,
            end.checked_sub(1),
            alignment,
            ranges: BTreeMap::new(),
        };
```

Reply:
The key point is to remove the error handle path in allocator.ranges.insert(base.checked_add(size)?, 0);

Reply:
Yep, got the point. I mean we need minus 1 again to store the real end number. But we can directly use unchecked_sub(1) since size>=0.
```
        let bound = base.checked_add(size)?;
        let end = bound.unchecked_sub(1); 
        let mut allocator = AddressAllocator {
            base,
            end,
            alignment,
            ranges: BTreeMap::new(),
        };
         allocator.ranges.insert(bound, 0);
```